polynote-kernel/src/main/scala/polynote/kernel/lang/LanguageKernel.scala

```diff
@@ -2,60 +2,42 @@ package polynote.kernel.lang
 
 import java.io.File
 
-import fs2.concurrent.{Enqueue, Queue}
+import fs2.Stream
 import polynote.kernel._
-import polynote.kernel.util.{Publish, RuntimeSymbolTable, SymbolDecl}
+import polynote.kernel.context.{GlobalInfo, RuntimeContext, SymbolDecl}
 
 /**
   * The LanguageKernel runs code in a given language.
   */
-trait LanguageKernel[F[_]] {
-
-  // LanguageKernel is expected to have a reference to the shared runtime symbol table of a notebook
-  val symbolTable: RuntimeSymbolTable
-
-  final type Decl = SymbolDecl[F, symbolTable.global.type]
-
-  def predefCode: Option[String]
+trait LanguageKernel[F[_], G <: GlobalInfo] {
 
   /**
     * Run the given code.
     *
-    * @param cell           The identifier string of the cell for the code being run
-    * @param visibleSymbols A list of symbols defined in cells "before" the given code, which are visible to it
-    * @param previousCells  The identifier strings of the cells "before" this code, for stateful language kernels
-    * @param code           The code string to run
-    * @param out            A [[Queue]] which should receive [[Result]] value(s) describing the output(s) from running the code
-    * @param statusUpdates  A [[Publish]] which should receive [[KernelStatusUpdate]]s describing changes in runtime status
-    *                       from running the cell, such as newly defined symbols, subtask progress, etc
+    * @param cell            The identifier string of the cell for the code being run
+    * @param runtimeContext  A View into the runtime context for this particular cell
+    * @param code            The code string to run
+    * @return                A Tuple of Streams: (results, new runtime symbols, maybe a cell return value)
     */
   def runCode(
     cell: String,
-    visibleSymbols: Seq[Decl],
-    previousCells: Seq[String],
-    code: String,
-    out: Enqueue[F, Result],
-    statusUpdates: Publish[F, KernelStatusUpdate]
-  ): F[Unit]
+    runtimeContext: RuntimeContext[G], // TODO: this could just be the parent cell id...
+    code: String
+  ): F[(Stream[F, Result], F[RuntimeContext[G]])]
```

Comment:
I'm not sure I like `runCode` returning a `RuntimeContext`. I feel like it should return a `Stream` and that the caller should worry about managing the `RuntimeContext`.

I guess it could return a `F[Stream[F, Either[Result, G#RuntimeValue]]]` (or we could make an ADT to encompass results and symbols). Then it would only have to return one stream. Caller would evaluate the stream and collect relevant stuff into a `RuntimeContext` frame.